internal/pkg/otel/configtranslate/otelconfig.go

```diff
@@ -0,0 +1,368 @@
+	if unit.Type == client.UnitTypeInput {
+		return nil, fmt.Errorf("unit type is an input, expected output: %v", unit)
+	}
+	configTranslationFunc, ok := configTranslationFuncForExporter[exporterType]
+	if !ok {
+		return nil, fmt.Errorf("no config translation function for exporter type: %s", exporterType)
+	}
+	// we'd like to use the same exporter for all outputs with the same name, so we parse out the name for the unit id
+	// these will be deduplicated by the configuration merging process at the end
+	outputName := strings.TrimPrefix(unit.ID, inputType+"-") // TODO: Use a more structured approach here
+	exporterId := getExporterID(exporterType, outputName)
+
+	// translate the configuration
+	unitConfigMap := unit.Config.GetSource().AsMap() // this is what beats do in libbeat/management/generate.go
+	outputCfgC, err := config.NewConfigFrom(unitConfigMap)
+	if err != nil {
+		return nil, fmt.Errorf("error translating config for output: %s, unit: %s, error: %w", outputName, unit.ID, err)
+	}
+	exporterConfig, err := configTranslationFunc(outputCfgC)
+	if err != nil {
+		return nil, fmt.Errorf("error translating config for output: %s, unit: %s, error: %w", outputName, unit.ID, err)
+	}
+
+	exportersCfg := map[string]any{
+		exporterId.String(): exporterConfig,
+	}
+
+	return exportersCfg, nil
+}
+
+// getInputsForUnit returns the beat inputs for a unit. These can directly be plugged into a beats receiver config.
+// It mainly calls a conversion function from the control protocol client.
+func getInputsForUnit(unit component.Unit, info info.Agent, defaultDataStreamType string) ([]map[string]any, error) {
+	agentInfo := &client.AgentInfo{
+		ID:           info.AgentID(),
+		Version:      info.Version(),
+		Snapshot:     info.Snapshot(),
+		ManagedMode:  runtime.ProtoAgentMode(info),
+		Unprivileged: info.Unprivileged(),
+	}
+	inputs, err := management.CreateInputsFromStreams(unit.Config, defaultDataStreamType, agentInfo)
+	if err != nil {
+		return nil, err
+	}
+	return inputs, nil
+}
+
+// getDefaultDatastreamTypeForComponent returns the default datastream type for a given component.
+// This is needed to translate from the agent policy config format to the beats config format.
+func getDefaultDatastreamTypeForComponent(comp *component.Component) (string, error) {
+	beatName := getBeatNameForComponent(comp)
+	switch beatName {
+	case "filebeat":
+		return "logs", nil
+	case "metricbeat":
+		return "metrics", nil
+	default:
+		return "", fmt.Errorf("input type not supported by Otel: %s", comp.InputType)
+	}
+}
+
+// translateEsOutputToExporter translates an elasticsearch output configuration to an elasticsearch exporter configuration.
+func translateEsOutputToExporter(cfg *config.C) (map[string]any, error) {
+	esConfig, err := elasticsearchtranslate.ToOTelConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+	// we want to use dynamic indexing
+	esConfig["logs_dynamic_index"] = map[string]any{"enabled": true}
```

Comment:
We may want to provide a way to supplement the ES exporter configuration directly from the elastic-agent.yml configuration. Inevitably we will want to change some part of it, and we may be annoyed waiting 12h for the snapshot build job to propagate it everywhere.

Reply:
Filed an issue here: https://github.com/elastic/ingest-dev/issues/5128.